### DISCUSSION TITLE
Unnecessary unique constraint on primary key

### DISCUSSION THREAD

**Discussion by fruttasecca** [sqlalchemy/sqlalchemy#5887] [Usage Questions] [answered]
Hi there, I have been noticing that the sqlalchemy ORM layer issues an unnecessary unique constraint to columns that are primary keys. I haven't had any luck finding out why this is the case and when it is desirable. 

```python
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer

conn = "postgresql://postgres@<set me>/<set me>"
engine = create_engine(conn, echo=True)
Base = declarative_base()

class User(Base):
     __tablename__ = 'users'

     id = Column(Integer, primary_key=True, unique=True)
     def __repr__(self):
        return str(self.id)

Base.metadata.create_all(engine)
```

This is the resulting command:
```sql
CREATE TABLE users (
	id SERIAL NOT NULL, 
	PRIMARY KEY (id), 
	UNIQUE (id)
)
```

**Versions**
 - OS: Ubuntu 18.04
 - Python: 3.7.6
 - SQLAlchemy: 1.3.13
 - Database: PostgreSQL 13.1 (Debian 13.1-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
 - DBAPI: psycopg2, 2.8.6

Thank you

**Comment by CaselIT** [comment, marked as answer]
Hi,

This happens because you are also specifying `unique=True` in the column definition. If you remove it it should no longer render the unique constraints.
As to why it happens, I think it's because the primary key and unique constraints are considered separately and for multi-column primary key the two would not be equivalent.

**Comment by fruttasecca** [comment]
Hi CaselIT,

>This happens because you are also specifying unique=True in the column definition. If you remove it it should no longer render the unique constraints.

Indeed, I should have clarified that the `unique` argument was intentionally put there in the column declaration to trigger the example behavior, my bad.

>As to why it happens, I think it's because the primary key and unique constraints are considered separately and for multi-column primary key the two would not be equivalent.

The why is what I am mostly interested in, I have reached the same conclusion, although I would like to understand if such behavior (single column PK getting a unique constraint) is intentional or the result of the implementation,  if the former, I would like to understand what are the implications.

**Reply by CaselIT** [reply]
I think it's just how it's implemented, and I don't think sqlalchemy should decide for the user if an unique constraint is or is not required for a particular column since inevitably there will be false positives/negatives.
`unique=True` expresses a fairly clear intent by the user to have a unique constraint that sqlalchemy should just apply.

Regarding implication I think it depends on the particular db, but I would hazard that it does not behave in any particular way compared to an unique constraint on a non pk column

**Comment by zzzeek** [comment]
it is absolutely valid for a unique constraint to be placed on a column that is also part of a primary key constraint.   in the case of a single-column, i.e. non-composite primary key, it is indeed redundant.   but it's not SQLAlchemy's job to second-guess explicit intent.    A valid use case would be someone trying to mimic the exact implementation of a legacy schema in order to duplicate some database-specific behavior, or perhaps there's a database system where a "primary key column" is actually composite against something else, such as an inherited table of some kind.      in any case I think if "unique=True" decided to do nothing in certain cases that would likely be considered to be more surprising.

the docs for unique=True / index=True are lacking in a few areas so i will push some more accurate language and important things to know about (e.g. naming conventions).

**Reply by fruttasecca** [reply]
Hi zzzeek, 

Thanks for the clarification, much appreciated.

**Reply by s2t2** [reply]
@zzzeek could you please clarify whether we need to separately specify a uniqueness constraint for a primary key (non-composite), or if that uniqueness constraint is already implemented by the primary key designation?

**Reply by zzzeek** [reply]
primary key constraints by definition constrain to uniqueness whether or not they are composite.
